bin/transform.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python
+
+import argparse
+import pathlib
+import pandas
+import numpy
+
+from dateutil.relativedelta import relativedelta
+from datetime import datetime
+
+# Column headers:
+SAMPLE_HEADER = "sample"
+SAMPLE_NAME_HEADER = "sample_name"
+AGE_HEADER = "age"
+VALID_HEADER_EXTENSION = "_valid"
+ERROR_HEADER_EXTENSION = "_error"
+
+# Column indices:
+DATE_1_INDEX = 2
+DATE_2_INDEX = 3
+
+# Transformations:
+LOCK = "lock"
+AGE = "age"
+
+# Other:
+AGE_THRESHOLD = 2 # Ages less than this will include a decimal component.
+DAYS_IN_YEAR = 365.0
+
+def remove_empty_columns(metadata):
+    metadata.dropna(axis="columns", how="all", inplace=True)
+
+def lock(metadata):
+    metadata_readable = metadata.copy(deep=True)
+    metadata_irida = metadata.drop(columns=[SAMPLE_NAME_HEADER], inplace=False)
+
+    return metadata_readable, metadata_irida
+
+def format_age(age):
+    if age < AGE_THRESHOLD:
+        formatted_age = "{:.4f}".format(age)
+    else:
+        formatted_age = "{:.0f}".format(age)
+
+    return formatted_age
+
+def calculate_age(row):
+    pattern = "%Y-%m-%d"
+    age = numpy.nan
+    # numpy.nan, not pandas.NA because numpy.nan is treated as a float.
+    # Otherwise, there's a risk of mixing age floats with pandas.NA and having
+    # the column be treated as an object column, which will prevent
+    # .to_csv(..., float_format=format_age) from working.
+    age_valid = False
+    age_error = "Unable to calculate age."
+
+    # Is a date missing?
+    if pandas.isnull(row.iloc[DATE_1_INDEX]) or pandas.isnull(row.iloc[DATE_2_INDEX]):
+        age = numpy.nan
+        age_valid = False
+        age_error = "At least one of the dates is missing."
+
+        return pandas.Series([age, age_valid, age_error])
+
+    date_1_string = row.iloc[DATE_1_INDEX]
+    date_2_string = row.iloc[DATE_2_INDEX]
+
+    # Are the dates in the correct format?
+    try:
+        date_1 = datetime.strptime(date_1_string, pattern)
+        date_2 = datetime.strptime(date_2_string, pattern)
+
+    except ValueError:
+        age = numpy.nan
+        age_valid = False
+        age_error = "The date format does not match the expected format (YYYY-MM-DD)."
+
+        return pandas.Series([age, age_valid, age_error])
+
+    # Calculate the relative delta in calendar time:
+    relative_delta = relativedelta(date_2, date_1)
+
+    # Under age threshold, calculate as (days/days_in_year):
+    # Note: this is inaccurate, because how many days is a year?
+    if relative_delta.years < AGE_THRESHOLD:
+        time_delta = date_2 - date_1
+        age = time_delta.days / DAYS_IN_YEAR
+
+    # Age meets threshold, calculate as calendar years:
+    else:
+        age = relative_delta.years
```

Comment:
Related to the question of rounding, does this round to nearest year, or is it floor?

Reply:
```python
>>> date_1 = datetime.strptime("2000-01-01", "%Y-%m-%d")
>>> date_2 = datetime.strptime("2000-12-31", "%Y-%m-%d")
>>> relative_delta = relativedelta(date_2, date_1)
>>> relative_delta.years
0
```

Reply:
Thanks for confirming Eric. Thinking about this a bit more, I think `floor` makes sense for the age. Can always be changed later if requested.